From 1dd0be21431a8416a491cb61aac0575c9c14b71e Mon Sep 17 00:00:00 2001
From: Yuyang Huang <yuyanghuang@google.com>
Date: Wed, 08 Oct 2025 17:40:29 -0700
Subject: [PATCH] [M141] Skip internal load flag check for multi-network CCT

Original change's description:
> Skip internal load flag check for multi-network CCT
> 
> In the multi-network CCT workflow, for subresource query a nested
> `CorsURLLoaderFactory` is created to run the same request on a specific
> network. This causes `IsValidRequest()` to be invoked twice. The first
> invocation performs the authoritative validation; later passes may add
> internal flags to the request. Re-validating the internal load flag is
> not set at this point is unnecessary and can falsely reject otherwise
> valid requests.
> 
> Bug: b:446461400
> Change-Id: I8a50e362ef27d93840aea9948e070fee78421bcb
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7007747
> Reviewed-by: Stefano Duo <stefanoduo@google.com>
> Commit-Queue: Yuyang Huang <yuyanghuang@google.com>
> Reviewed-by: Adam Rice <ricea@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1526030}

(cherry picked from commit b4624bd29e4bc6fea31adb0426b65f95ecf9585d)

Bug: 449877452,b:446461400
Change-Id: I8a50e362ef27d93840aea9948e070fee78421bcb
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7022871
Bot-Commit: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: Chrome Cherry Picker <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Commit-Queue: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/7390@{#2201}
Cr-Branched-From: d481efce5eb300acbb896686676ebd0352a6f1db-refs/heads/main@{#1509326}
---

--- a/services/network/cors/cors_url_loader_factory.cc
+++ b/services/network/cors/cors_url_loader_factory.cc
@@ -544,6 +544,11 @@
   return true;
 }
 
+bool CorsURLLoaderFactory::IsMultiNetworkCCTWorkFlow() const {
+  return context_->url_request_context()->bound_network() !=
+         net::handles::kInvalidNetworkHandle;
+}
+
 bool CorsURLLoaderFactory::IsCorsPreflighLoadOptionAllowed() const {
   // kURLLoadOptionAsCorsPreflight is set by CorsURLLoader itself, when
   // starting a request, if CORS preflight request is needed.
@@ -578,9 +583,7 @@
   // a valid network. So, given that this config is security critical, it's best
   // to "peek into implementation details" rather than granting this exception
   // to a bigger group.
-  return allow_external_preflights_for_testing_ ||
-         context_->url_request_context()->bound_network() !=
-             net::handles::kInvalidNetworkHandle;
+  return allow_external_preflights_for_testing_ || IsMultiNetworkCCTWorkFlow();
 }
 
 bool CorsURLLoaderFactory::IsValidRequest(const ResourceRequest& request,
@@ -607,8 +610,24 @@
   if (request.load_flags &
       (net::LOAD_CAN_USE_SHARED_DICTIONARY |
        net::LOAD_DISABLE_SHARED_DICTIONARY_AFTER_CROSS_ORIGIN_REDIRECT)) {
-    mojo::ReportBadMessage("CorsURLLoaderFactory: Internal load flag received");
-    return false;
+    // In the multi-network CCT workflow, when fetching the subresource, we
+    // create a nested CorsURLLoaderFactory to run the same request on a
+    // specific network. That causes this check to be invoked twice: the first
+    // invocation performs the authoritative validation, and later passes may
+    // add internal flags to the request. Re-validating here is unnecessary and
+    // can falsely reject otherwise valid requests (See
+    // CorsURLLoaderFactory::IsCorsPreflighLoadOptionAllowed for the rational).
+    // Note: Skipping this check is appropriate for subresource requests.
+    // For main page loads or navigation requests, ideally the check should
+    // still be performed as there is no nesting. However, since this is only
+    // a sanity check, skipping it here is safe.
+    // TODO(crbug.com/449098586): refactor to get rid of nested
+    // CorsURLLoaderFactory to make the design clearer.
+    if (!IsMultiNetworkCCTWorkFlow()) {
+      mojo::ReportBadMessage(
+          "CorsURLLoaderFactory: Internal load flag received");
+      return false;
+    }
   }
 
   // Check if this is an untrusted factory being provided parameters that should
--- a/services/network/cors/cors_url_loader_factory.h
+++ b/services/network/cors/cors_url_loader_factory.h
@@ -160,6 +160,9 @@
   mojo::PendingRemote<mojom::DevToolsObserver> GetDevToolsObserver(
       ResourceRequest& resource_request) const;
 
+  // Returns whether this factory is used in the multi-network CCT workflow.
+  bool IsMultiNetworkCCTWorkFlow() const;
+
   template <class T>
   void OnLoaderCreated(
       std::unique_ptr<T> loader,
